flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java

```diff
@@ -63,11 +70,19 @@
 
 	private boolean isShutdown;
 
-	public NetworkEnvironment(NetworkEnvironmentConfiguration config, TaskEventPublisher taskEventPublisher) {
+	public NetworkEnvironment(
```

Comment:
From dependency injection perspective, could we add a static factory method here which would create all dependency components, like networkBufferPool, connectionManager, doing metrics registration etc? Then it could just call the private constructor passing all configured dependencies. I know we do not use this constructor at the moment but it would help to keep code cleaner in long term. At the end `ShuffleMaster.createShuffleService(flinkConfig)` could merge parsing `NetworkEnvironmentConfiguration` and creation of dependency components before creating `NetworkEnvironment` as `ShuffleService`.

Reply:
Yes, I also considered this idea before, but not sure when to bring this factory. I could try to do it in this PR.

Reply:
I guess your mentioned factory is similar with the `Builder` function to create every internal components or pass them from outsides as in test usages.

I thought of this issue from the motivation of tests aspect, but there seems only one `TaskEventDispatcher` component, so I have not focused on it before. Now I add the new component `ComponentMetricGroup`,  it seems necessary. 

My suggestion is providing the simple build method atm for avoiding creating these two components in related tests. The final form of factory is related to `NetworkEnvironment#fromConfiguration(flink configuration)` and `ShuffleMaster.createShuffleService(flinkConfig)`, so it might be better to consider them together when working the PR for `ShuffleMaster.createShuffleService(flinkConfig)`. What do you think?

Reply:
agreed, it is actually more related to the below comment about `NetworkEnvironment` creation refactoring in `TaskManagerServices`.